experimental/parser/parse_state.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"github.com/bufbuild/protocompile/experimental/ast"
+	"github.com/bufbuild/protocompile/experimental/internal/taxa"
+	"github.com/bufbuild/protocompile/experimental/report"
+	"github.com/bufbuild/protocompile/experimental/token"
+)
+
+// lexer is a Protobuf parser.
+type parser struct {
+	ast.Context
+	*ast.Nodes
+	*report.Report
+}
+
+// parsePunct attempts to unconditionally parse some punctuation.
+//
+// If the wrong token is encountered, it DOES NOT consume the token, returning a nil
+// token instead. Returns a diagnostic on failure.
+func (p *parser) Punct(c *token.Cursor, want string, where taxa.Place) (token.Token, report.Diagnose) {
+	next := c.Peek()
+	if next.Text() == want {
+		return c.Pop(), nil
+	}
+
+	wanted := taxa.NewSet(taxa.Punct(want, false))
+	if next.Nil() {
+		tok, span := c.JustAfter()
+		err := errUnexpected{
+			what:  span,
+			where: where,
+			want:  wanted,
+			got:   taxa.EOF,
+		}
+		if !tok.Nil() {
+			err.got = taxa.Classify(tok)
+		}
+
+		return token.Nil, err
+	}
+
+	return token.Nil, errUnexpected{
+		what:  next,
+		where: where,
+		want:  wanted,
+	}
+}
+
+// commas returns an iterator over a comma-delimited list of things out of cursor.
+// This automatically handles various corner-cases around commas that occur throughout the
+// grammar.
+//
+// This will completely drain cursor, unless the parse function returns false, which signals
+// that the end of the list has been reached.
```

Comment:
This isn't true. There's an early return if `commasRequired` is true but no comma is found. Seems like that should instead add a diagnostic instead of breaking out of loop.

Reply:
No, adding a diagnostic isn't the right thing to do here: we want parsing to continue to the next step if no comma was present, since it signals the end of the list.

For example, consider

```proto
reserved 1, 2, 3;
```

Upon seeing the ; after 3, commas will return. I've updated the comment to reflect this.

Reply:
Perhaps this should be a flag? Because when it's a list inside of `{...}` or `[...]`, we should try to consume everything. And, honestly, even in the example given, we shouldn't bail at the first non-comma but perhaps only at a semi-colon, so that we can diagnose missing commas in `reserved 1 2 3;`.

Reply:
I think I need to refactor commas in *this* PR...

Reply:
Done.